src/literacy.rs

```diff
@@ -0,0 +1,263 @@
+// Bitcoin Hashes Library
+// Written in 2021 by
+//   The rust-bitcoin developers.
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # Literacy traits
+//!
+//! literacy: "the ability to [Read] and [Write]."
+//!
+//! * With `std` enabled, traits are automatically implemented for `std::io::{Read, Write}`
+//! * Without `std` but using feature `use-core2`, they are implemented for `core2::io::{Read, Write}`
+//! * Without neither `std` and nor `core2` default implementation `impl Read for &[u8]`
+//!   and `impl Write for ::alloc::vec::Vec<u8>` are provided
+//!
+
+/// The Read trait allows for reading bytes from a source.
+pub trait Read {
+    /// The error type returned in Result
+    type Error;
+    /// The type to implement limited reads
+    type Take: Read;
+    /// see [std::io::Read::read]
+    fn read(&mut self, buf: &mut [u8]) -> ::core::result::Result<usize, Self::Error>;
+    /// see [std::io::Read::read_exact]
+    fn read_exact(&mut self, buf: &mut [u8]) -> ::core::result::Result<(), Self::Error>;
+    /// see [std::io::Read::take]
+    fn take(self, limit: u64) -> Self::Take;
```

Comment:
This doesn't seem to quite work with the `&mut` pattern:

```
    fn consensus_decode<D: io::Read>(d: &mut D) -> Result<Self, encode::Error> {
        let mut d = d.take(MAX_VEC_SIZE as u64);
```

gives:

```
error[E0507]: cannot move out of `*d` which is behind a mutable reference

    let mut d = d.take(MAX_VEC_SIZE as u64);
                      ^ move occurs because `*d` has type `D`, which does not implement the `Copy` trait
```

Reply:
maybe removing `take` from the trait and using something like the CappedRead in https://github.com/RCasatta/rust-bitcoin/commit/24015e8ed17a4d880311247cba2b58b53deaa890 ?

Reply:
I'll give this a try.

Reply:
Seems to work fine, let's take out the `take`.

Reply:
take removed in c6d4ca48fd56189a1c20b08c8fd81786b3b65d74